libsplinter/src/service/network/interconnect/mod.rs

```diff
@@ -13,3 +13,707 @@
+    ///
+    /// # Arguments
+    ///
+    /// * `service_msg_dispatcher` - a Dispatcher that is loaded with the Handlers for
+    ///    ComponentMessageTypes
+    pub fn with_service_msg_dispatcher(
+        mut self,
+        service_msg_dispatcher: Dispatcher<ComponentMessageType, ConnectionId>,
+    ) -> Self {
+        self.service_msg_dispatcher = Some(service_msg_dispatcher);
+        self
+    }
+
+    /// Build the ServiceInterconnect. This function will build the dispatch loop for network message
+    /// types and start up threads to send and recv messages from the services.
+    ///
+    /// Returns the ServiceInterconnect object that can be used to get network message senders and
+    /// shutdown message threads.
+    pub fn build(&mut self) -> Result<ServiceInterconnect, ServiceInterconnectError> {
+        let (dispatched_sender, dispatched_receiver) = channel();
+
+        let service_lookup_provider = self.service_lookup_provider.take().ok_or_else(|| {
+            ServiceInterconnectError::StartUpError("Service lookup provider missing".to_string())
+        })?;
+
+        let mut service_msg_dispatcher = self.service_msg_dispatcher.take().ok_or_else(|| {
+            ServiceInterconnectError::StartUpError("Network dispatcher missing".to_string())
+        })?;
+        service_msg_dispatcher.set_network_sender(ServiceInterconnectMessageSender::new(
+            dispatched_sender.clone(),
+        ));
+        let component_dispatch_loop = DispatchLoopBuilder::new()
+            .with_dispatcher(service_msg_dispatcher)
+            .build()
+            .map_err(|err| {
+                ServiceInterconnectError::StartUpError(format!(
+                    "Unable to start network dispatch loop: {}",
+                    err
+                ))
+            })?;
+        let service_msg_dispatcher_sender = component_dispatch_loop.new_dispatcher_sender();
+        let service_msg_dispatcher_shutdown = component_dispatch_loop.shutdown_signaler();
+
+        // start receiver loop
+        let message_receiver = self.message_receiver.take().ok_or_else(|| {
+            ServiceInterconnectError::StartUpError("Message receiver missing".to_string())
+        })?;
+
+        let recv_service_lookup = service_lookup_provider.service_lookup();
+        let recv_join_handle = thread::Builder::new()
+            .name("ServiceInterconnect Receiver".into())
+            .spawn(move || {
+                if let Err(err) = run_recv_loop(
+                    &*recv_service_lookup,
+                    message_receiver,
+                    service_msg_dispatcher_sender,
+                ) {
+                    error!("Shutting down service interconnect recevier: {}", err);
+                }
+            })
+            .map_err(|err| {
+                ServiceInterconnectError::StartUpError(format!(
+                    "Unable to start ServiceInterconnect receiver thread {}",
```

Comment:
Might have missed a semicolon before the error (same for sender thread below)

Reply:
Fixed